include/oneapi/mkl/dft/backward.hpp

```diff
@@ -26,51 +26,137 @@
 #include <CL/sycl.hpp>
 #endif
 
+#include "detail/types_impl.hpp"
+
 namespace oneapi::mkl::dft {
 //Buffer version
 
 //In-place transform
 template <typename descriptor_type, typename data_type>
-void compute_backward(descriptor_type &desc, sycl::buffer<data_type, 1> &inout);
+void compute_backward(descriptor_type &desc, sycl::buffer<data_type, 1> &inout) {
+    static_assert(detail::valid_compute_arg<descriptor_type, data_type>::value,
+                  "unexpected type for data_type");
+
+    using fwd_type = typename detail::descriptor_info<descriptor_type>::forward_type;
+    auto type_corrected_inout = inout.template reinterpret<fwd_type, 1>(
+        detail::reinterpret_range<data_type, fwd_type>(inout.size()));
+    get_commit(desc)->backward_ip_cc(desc, type_corrected_inout);
+}
 
 //In-place transform, using config_param::COMPLEX_STORAGE=config_value::REAL_REAL data format
 template <typename descriptor_type, typename data_type>
 void compute_backward(descriptor_type &desc, sycl::buffer<data_type, 1> &inout_re,
-                      sycl::buffer<data_type, 1> &inout_im);
+                      sycl::buffer<data_type, 1> &inout_im) {
+    static_assert(detail::valid_compute_arg<descriptor_type, data_type>::value,
+                  "unexpected type for data_type");
+
+    using scalar_type = typename detail::descriptor_info<descriptor_type>::scalar_type;
+    auto type_corrected_inout_re = inout_re.template reinterpret<scalar_type, 1>(
+        detail::reinterpret_range<data_type, scalar_type>(inout_re.size()));
+    auto type_corrected_inout_im = inout_im.template reinterpret<scalar_type, 1>(
+        detail::reinterpret_range<data_type, scalar_type>(inout_im.size()));
+    get_commit(desc)->backward_ip_rr(desc, type_corrected_inout_re, type_corrected_inout_im);
+}
 
 //Out-of-place transform
 template <typename descriptor_type, typename input_type, typename output_type>
 void compute_backward(descriptor_type &desc, sycl::buffer<input_type, 1> &in,
-                      sycl::buffer<output_type, 1> &out);
+                      sycl::buffer<output_type, 1> &out) {
+    static_assert(detail::valid_compute_arg<descriptor_type, input_type>::value,
+                  "unexpected type for input_type");
+    static_assert(detail::valid_compute_arg<descriptor_type, output_type>::value,
+                  "unexpected type for output_type");
+
+    using fwd_type = typename detail::descriptor_info<descriptor_type>::forward_type;
+    using bwd_type = typename detail::descriptor_info<descriptor_type>::backward_type;
+    auto type_corrected_in = in.template reinterpret<bwd_type, 1>(
```

Comment:
whats the benifit of converting the `input_type` to 'scalar_type' which is strictly only a 'float' or 'double'.Is it because you want to supprot not only scalar and complex in/out_type but sycl::float2/4 double2/4, or some nuance of the rocFFT/cuFFT ?

Reply:
I'm not completely sure I understand your question but I'll give you some info that might clarify things.
The user facing functions `compute_forward/compute_backward` currently have no restrictions on the types of `input_type` and `output_type` (though I have cheekily added some sensible ones in a static assert!). Where the functions are not user facing, I've made the parameter types purely dependent on the descriptor type, with (hopefully) sensible types chosen in the [`descriptor_info`](https://github.com/oneapi-src/oneMKL/blob/4905064a608aa89e47e152e0d0480239b5f06251/include/oneapi/mkl/dft/detail/types_impl.hpp#L61) struct.
The casts that happen in `include/oneapi/mkl/dft/backward.hpp` and `include/oneapi/mkl/dft/forward.hpp` are there to reinterpret the type that the user has chosen to the internal types.

Reply:
Well what I mean is reinterpret_type converts any valid type that passes the assertions just above to its fundamental type, i.e float or double, so my question is do we really need that ? Because in the sycl API we support float/double/complex<float/double> input, and in the C_API also we accept any of the former. So what exactly are we getting out of this reinterpretation?

Reply:
The type is not just reinterpreted to the scalar type, it is sometimes the `forward_type` or `backward_type` from `descriptor_info`.

I think it has a few benefits.
1. greater clarity about the expected data layout
2. reduced number of functions we need to expose from the library
3. reduced argument ambiguity (REAL_REAL in-place and COMPLEX_COMPLEX out-of-place have the same number of arguments, which can be confusing. With this system they at least take different arguments.)

Most importantly, the functions in `backward.hpp` call virtual functions on `commit_impl`, so those can't have a templated argument, so we either reinterpret or have ~100 virtual functions on the `commit_impl` class.